packages/typespec-client-generator-core/src/internal-utils.ts

```diff
@@ -252,12 +255,85 @@ export function getSdkTypeBaseHelper<TKind>(
   context: TCGCContext,
   type: Type,
   kind: TKind
-): DefaultSdkTypeBase<TKind> {
-  return {
+): [DefaultSdkTypeBase<TKind>, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  return diagnostics.wrap({
     __raw: type,
     deprecation: getDeprecationDetails(context.program, type)?.message,
     kind,
-  };
+    decorators: diagnostics.pipe(getTypeDecorators(context, type)),
+  });
+}
+
+export function getTypeDecorators(
+  context: TCGCContext,
+  type: Type
+): [Record<string, Record<string, any>>, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  const retval: Record<string, Record<string, any>> = {};
+  if ("decorators" in type) {
+    for (const decorator of type.decorators) {
+      // only process explicitly defined decorators
+      if (decorator.definition) {
+        const decoratorName = `${decorator.definition?.namespace ? getNamespaceFullName(decorator.definition.namespace) + "." : ""}${decorator.definition?.name}`;
+        // white list filtering
+        if (
+          !context.decoratorsWhiteList ||
+          !context.decoratorsWhiteList.some(
+            (x) => x === decoratorName || new RegExp(x).test(decoratorName)
+          )
+        ) {
+          continue;
+        }
+
+        retval[decoratorName] = {};
+        for (let i = 0; i < decorator.definition.parameters.length; i++) {
```

Comment:
consider to wrap this into a helper method getDecoratorWithArgs for readability

Reply:
Guess `for (const param of decorator.definition.parameters)` could be better.

Reply:
i need index to do cut.

Reply:
since this function is to get type decorators, i think such short part does need to be abstracted.

Reply:
I see